hudi-common/src/main/java/org/apache/hudi/common/table/log/HoodieMergedLogRecordScanner.java

```diff
@@ -153,9 +154,29 @@ protected void processNextRecord(HoodieRecord<? extends HoodieRecordPayload> hoo
   }
 
   @Override
-  protected void processNextDeletedKey(HoodieKey hoodieKey) {
-    records.put(hoodieKey.getRecordKey(), SpillableMapUtils.generateEmptyPayload(hoodieKey.getRecordKey(),
-        hoodieKey.getPartitionPath(), getPayloadClassFQN()));
+  protected void processNextDeletedRecord(DeleteRecord deleteRecord) {
+    String key = deleteRecord.getRecordKey();
+    if (records.containsKey(key)) {
+      // Merge and store the merged record. The ordering val is taken to decide whether the same key record
+      // should be deleted or be kept. The old record is kept only if the DELETE record has smaller ordering val.
+      // For same ordering values, uses the natural order(arrival time semantics).
+
+      HoodieRecord<? extends HoodieRecordPayload> oldRecord = records.get(key);
+      Comparable curOrderingVal = oldRecord.getData().getCombineValue();
+      Comparable deleteOrderingVal = deleteRecord.getOrderingVal();
+      // Checks the ordering value does not equal to 0
+      // because we use 0 as the default value which means natural order
+      boolean choosePrev = !deleteOrderingVal.equals(0)
+          && curOrderingVal.getClass() == deleteOrderingVal.getClass()
+          && curOrderingVal.compareTo(deleteOrderingVal) > 0;
```

Comment:
in retract stream, deleteOrderingVal and curOrderingVal are the same...
so the condition is `curOrderingVal.compareTo(deleteOrderingVal) >= 0` right？

Reply:
2： COW has the same issue？

Reply:
1. We take the process time semantics for DELETE and INSERT has the same event time.

2. COW has no issue because the DELETE was removed physically when generating new parquet file and in the in-memory deduplication we take the event time into account with invoking `#preCombine`.

Reply:
>We take the process time semantics if DELETE and INSERT has the same event time.

what does that mean? in retract stream, deleteOrderingVal and curOrderingVal are the same regardless of the job usage event time or process time.

Reply:
I guess there must be a primary key for hudi and there should be an ordering field, or we should totally use the process time sequence and the `DeleteRecord` should not be used, changelog mode is more suitable, this patch only fixes the default upsert behavior: the upstream should be a retract stream with event time.

Reply:
@vinothchandar Maybe we should really deprecate the delete block forever, imagine there are operations in one traction for a database:

INSERT DELETE INSERT (for a key, they all only have process time semantics, the linearizabiity)

The record is kept for a RDBMS but would lost if we do these operations in one commit in Hudi even if after this patch.

Reply:
Whats the concrete proposal to redo this? Can you write it down high level. I ll try to read this again and get the context.

Reply:
The concrete proposal original was to solve the data lost when there are: `DELETE`, `INSERT` events in one transaction/commit of hoodie table, if the events have correct event time field.

What the patch can not solve is: when the events have same event time field(processing sequence), the data would still be lost.

Reply:
>would lost if we do these operations in one commit in Hudi even if after this patch.

is that not a problem of how we define ordering field? are you taking about cases specific to Flink CDC?

Reply:
> are you taking about cases specific to Flink CDC

The Flink CDC output should have a event time field there, but now they are all processing time sequence.

For Debezium source it works well with this patch.